### PR TITLE
Add conversion to jni and java test

### PR DESCRIPTION
**Pull request by enzoc4**
- Fix a double free bug in the wallet recover jni function. [get_string](https://docs.rs/jni/0.16.0/jni/struct.JNIEnv.html#method.get_string) returns a RAII type, but `release_string_utf_chars` was called anyway on it (it only should be called on the result of `get_string_utf_chars`.
- Add the missing `Conversion` functions to the jni layer. 

(This was never triggered in my cordova plugin tests, for some reason. Either the gc didn't run, or something else swallows the error)

- Move the java functions that the plugin uses to `wallet-jni` and add a little java test there. I find it useful at least for developing (and already found a bug with it). Open to feedback on this part. I don't know if we can use something like this to for `wallet-android` @danielSanchezQ 

** WIP / TODO ** I still don't know how to really test the conversion. Also, missing the functions to get ignored values because I still don't know if I can return tuples or something (or if I have to create two functions)

### CODE REVIEW

**Comment by danielSanchezQ** [comment]
The testing in java is really helpful. We should probably do the same with Android. Actually, there are some tests in there (similar to those), the problem is that they need an Android runner (or I couldn't make it work without it). Now that I think of it, maybe we could simply run the tests with Kotlin. 
Also, maybe we can create a `.jar` with those java bindings for the release also? (Not sure if we actually want that, just dropping some ideas here).

File: bindings/wallet-jni/java/run_test.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+junit = "junit-4.13.jar"
+hamcrest = "hamcrest-core-1.3.jar"
+
+classpath = f".:{junit}:{hamcrest}"
+javafilesdir = Path("com/iohk/jormungandrwallet")
+
+
+def compile():
+    packageFiles = map(lambda path: str(path), (scriptdirectory /
```
Comment (danielSanchezQ) [inline bindings/wallet-jni/java/run_test.py:19]: nitpicking: `map(str, ...)`

File: bindings/wallet-jni/java/run_test.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+                                                Path("com/iohk/jormungandrwallet")).glob("*.java"))
+    out = subprocess.run([
+        "javac", "-cp", classpath, "WalletTest.java"] + list(packageFiles), cwd=scriptdirectory)
```
Comment (danielSanchezQ) [inline bindings/wallet-jni/java/run_test.py:22]: In `python3` you can also `["javac", "-cp", classpath, "WalletTest.java", *packageFiles]`, I think.

File: bindings/wallet-jni/src/lib.rs [hunk 3]
```diff
@@ -123,3 +121,101 @@ pub unsafe extern "system" fn Java_com_iohk_jormungandrwallet_Wallet_initialFund
...
+    let settings_ptr = settings as SettingsPtr;
+    let mut conversion_out = null_mut();
+    let result = wallet_convert(
+        wallet_ptr,
+        settings_ptr,
+        (&mut conversion_out) as *mut ConversionPtr,
+    );
+
+    if let Some(error) = result.error() {
+        env.throw(error.to_string());
```
Comment (danielSanchezQ) [inline bindings/wallet-jni/src/lib.rs:142]: I'm in a warnings hunting mode right now. Can we consume this result somehow?

File: bindings/wallet-jni/src/lib.rs [hunk 4]
```diff
@@ -123,3 +121,101 @@ pub unsafe extern "system" fn Java_com_iohk_jormungandrwallet_Wallet_initialFund
...
+        conversion,
+        index,
+        (&mut transaction_out) as *mut *const u8,
+        (&mut transaction_size) as *mut usize,
+    );
+
+    // TODO: maybe we can use a ByteBuffer or something here to avoid the copy
+    // (especially considering that there is already a requirement to call conversion's delete)
+    let array = env.new_byte_array(transaction_size as jint).expect("");
+    match result.error() {
```
Comment (danielSanchezQ) [inline bindings/wallet-jni/src/lib.rs:182]: Oh wow, this is weird, so we are mattching the result but still creates a warning? 🙉

File: bindings/wallet-jni/src/lib.rs [hunk 5]
```diff
@@ -123,3 +121,101 @@ pub unsafe extern "system" fn Java_com_iohk_jormungandrwallet_Wallet_initialFund
...
+    _: JClass,
+    wallet: WalletPtr,
+    value: jlong,
+    counter: jlong,
+) {
+    let wallet = wallet as WalletPtr;
+    let r = wallet_set_state(wallet, value as u64, counter as u32);
+
+    if let Some(error) = r.error() {
+        env.throw(error.to_string());
```
Comment (danielSanchezQ) [inline bindings/wallet-jni/src/lib.rs:219]: Same as above

File: bindings/wallet-jni/java/run_test.py (hunk 1 above)
Comment (enzoc4) [reply bindings/wallet-jni/java/run_test.py:19]: good point

File: bindings/wallet-jni/java/run_test.py (hunk 2 above)
Comment (enzoc4) [reply bindings/wallet-jni/java/run_test.py:22]: yes, it works. I like it, I didn't know that

File: bindings/wallet-jni/src/lib.rs (hunk 3 above)
Comment (enzoc4) [reply bindings/wallet-jni/src/lib.rs:142]: good catch, I actually let the warning on purpose to think about it later, and then I forgot. What @NicolasDP did for similar cases is this:
```rust 
let _ = env.throw(error.to_string());
```
so I guess I'll take the same route

File: bindings/wallet-jni/src/lib.rs (hunk 4 above)
Comment (enzoc4) [reply bindings/wallet-jni/src/lib.rs:182]: I think it's only about the `env.throw`

File: bindings/wallet-jni/java/run_test.py (hunk 1 above)
Comment (enzoc4) [reply bindings/wallet-jni/java/run_test.py]: I'm still not sure if I want to keep this as a python script, though. Maybe something like gradle would work here, but I don't have much experience with it.

Comment (danielSanchezQ) [reply bindings/wallet-jni/java/run_test.py]: Yeah, maybe gradle would be better from a java approach. I have not much experience myself too. I have seen in some posts that people use gradle to even call `cargo` so maybe at some point we can use it to pack it all together. I don't think we should lose too much time with that now so I guess the python script is enough. What do yo think @NicolasDP ?

**Comment by NicolasDP** [comment]
There is a bug fix with the transaction coming in #53

**Pull request merged**